compression/lz78.py

```diff
@@ -0,0 +1,127 @@
+"""
+Sources:
+https://en.wikipedia.org/wiki/LZ77_and_LZ78#LZ78
+"""
+
+from dataclasses import dataclass
+
+__version__ = "1.0"
+__author__ = "Ahmed Tamer"
+
+
+@dataclass
+class Token:
+    """
+    Dataclass representing pair called token consisting of the dictionary index
+    and a single character that follows the phrase in the dictionary.
+    """
+
+    index: int
+    char: str
+
+    def __repr__(self) -> str:
+        """
+        >>> token = Token(1, "c")
+        >>> repr(token)
+        '(1, c)'
+        >>> str(token)
+        '(1, c)'
+        """
+        return f"({self.index}, {self.char})"
+
+
+class LZ78Compressor:
+    """
+    Class containing compress and decompress methods using LZ78 compression algorithm.
+    """
+
+    def compress(self, text: str) -> list[Token]:
+        """
+        Compress the given string text using LZ78 compression algorithm.
+
+        Args:
+          text: string to be compressed
+
+        Returns:
+          output: the compressed text as a list of Tokens
+
+        Tests:
+          >>> lz78_compressor = LZ78Compressor()
+          >>> str(lz78_compressor.compress("ababcbababaa"))
+          '[(0, a), (0, b), (1, b), (0, c), (2, a), (5, b), (1, a)]'
+          >>> str(lz78_compressor.compress("aacaacabcabaaac"))
+          '[(0, a), (1, c), (1, a), (0, c), (1, b), (4, a), (0, b), (3, a)]'
+          >>> str(lz78_compressor.compress(""))
+          '[]'
+          >>> str(lz78_compressor.compress([]))
```

Comment:
The [] and {} tests should raise an Exception to warn the user not to provide garbage input.